From cbf97d4b00e8fb9d46eb4c53bf64073871d7ce87 Mon Sep 17 00:00:00 2001
From: Wim Taymans <wtaymans@redhat.com>
Date: Mon, 7 Aug 2023 19:58:20 +0200
Subject: [PATCH] jack: handle node.always-process = false jack nodes

Node that have the node.always-process = false property do not conform
to the jack API because they will be suspended even when they don't
inactivate themselves. Don't hide the ports for those clients when
inactive.

Fixes #3416
---
 pipewire-jack/src/pipewire-jack.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/pipewire-jack/src/pipewire-jack.c b/pipewire-jack/src/pipewire-jack.c
index 7547d6ffb..051eb1675 100644
--- a/pipewire-jack/src/pipewire-jack.c
+++ b/pipewire-jack/src/pipewire-jack.c
@@ -3075,11 +3075,18 @@ static void node_info(void *data, const struct pw_node_info *info)
 {
 	struct object *n = data;
 	struct client *c = n->client;
+	const char *str;
+
+	if (info->change_mask & PW_NODE_CHANGE_MASK_PROPS) {
+		str = spa_dict_lookup(info->props, PW_KEY_NODE_ALWAYS_PROCESS);
+		n->node.is_jack = str ? spa_atob(str) : false;
+	}
 
-	pw_log_info("DSP node %d state change %s", info->id,
-			pw_node_state_as_string(info->state));
+	n->node.is_running = !n->node.is_jack || (info->state == PW_NODE_STATE_RUNNING);
 
-	n->node.is_running = (info->state == PW_NODE_STATE_RUNNING);
+	pw_log_debug("DSP node %d %08"PRIx64" jack:%u state change %s running:%d", info->id,
+			info->change_mask, n->node.is_jack,
+			pw_node_state_as_string(info->state), n->node.is_running);
 
 	if (info->change_mask & PW_NODE_CHANGE_MASK_STATE) {
 		struct object *p;
-- 
GitLab

